File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/CommentService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Comment;
import at.ac.tuwien.sepm.groupphase.backend.exception.CommentNotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.DeckNotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.AuthenticationRequiredException;
import at.ac.tuwien.sepm.groupphase.backend.exception.InsufficientAuthorizationException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface CommentService {

    /**
     * Add a new comment to an existing deck
     *
     * @param deckId id of the deck where it will be added
     * @param comment data of the new comment
     * @return created comment
     * @throws DeckNotFoundException if no deck with this id exists
     * @throws AuthenticationRequiredException if no authenticated user could be found
     */
    Comment addCommentToDeck(Long deckId, Comment comment);

    /**
     * Find a single comment by id
     *
     * @param commentId id of the comment
     * @return the comment
     * @throws CommentNotFoundException if no comment with this id exists
     */
    Comment findOneOrThrow(Long commentId);

    /**
     * Get page of comments for a deck
     *
     * @param deckId of the deck
     * @param pageable pagination for the query
     * @return page of comments of this deck
     * @throws DeckNotFoundException if no deck with this id exists
     */
    Page<Comment> findCommentsByDeckId(Long deckId, Pageable pageable);

    /**
     * Update an existing comment
     *
     * @param commentId id of the comment to edit
     * @param comment new data of the comment
     * @return edited comment
     * @throws CommentNotFoundException if no comment with this id exists
     * @throws AuthenticationRequiredException if no authenticated user could be found
     * @throws InsufficientAuthorizationException if the comment does not belong to the logged in user
     */
    Comment editComment(Long commentId, Comment comment);

    /**
     * Delete a comment
     *
     * @param commentId id of the comment to delete
     * @throws CommentNotFoundException if no comment this this id exists
     * @throws AuthenticationRequiredException if no authenticated user could be found
     * @throws InsufficientAuthorizationException if the comment does not belong to the logged in user and user is no admin
     */
    void deleteComment(Long commentId);
}

File: frontend/src/app/services/comment.service.ts
import { Injectable } from '@angular/core';
import { CommentSimple } from '../dtos/commentSimple';
import { HttpClient } from '@angular/common/http';
import { Globals } from '../global/globals';
import { ErrorHandlerService } from './error-handler.service';
import { AuthService } from './auth.service';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { Page } from '../dtos/page';
import { Pageable } from '../dtos/pageable';

@Injectable({
  providedIn: 'root'
})
export class CommentService {

  private deckBaseUri = this.globals.backendUri + '/decks'
  private commentBaseUri = this.globals.backendUri + '/comments'

  constructor(private httpClient: HttpClient, private globals: Globals, private errorHandler: ErrorHandlerService, private authService: AuthService) {
  }

  /**
   * Add a comment to a deck
   * 
   * @param deckId deck where the comment should be added
   * @param message comment message to be added
   * @return created comment
   */
  addCommentToDeck(deckId: number, message: string): Observable<CommentSimple> {
    const dto = { message }
    return this.httpClient.post<CommentSimple>(`${this.deckBaseUri}/${deckId}/comments`, dto)
      .pipe(tap(null, this.errorHandler.handleError('Could not create comment')))
  }

  /**
   * Edit an existing comment
   * 
   * @param commentId id of the comment to edit
   * @param message new comment message
   * @return updated comment
   */
  editComment(commentId: number, message: string): Observable<CommentSimple> {
    const dto = { message }
    return this.httpClient.put<CommentSimple>(`${this.commentBaseUri}/${commentId}`, dto)
      .pipe(tap(null, this.errorHandler.handleError('Could not edit comment')))
  }

  /**
   * Find page of comments of a deck
   * 
   * @param deckId deck which should be searched for comments
   * @param pageable pagination config
   * @return page of comments sorted by newest first
   */
  findByDeckId(deckId: number, pageable: Pageable): Observable<Page<CommentSimple>> {
    return this.httpClient.get<Page<CommentSimple>>(`${this.deckBaseUri}/${deckId}/comments`, { params: pageable.toHttpParams() })
      .pipe(tap(null, this.errorHandler.handleError('Could not fetch comments')))
  }

  /**
   * Find a single comment
   * 
   * @param commentId id of the comment
   * @return comment
   */
  findOne(commentId: number): Observable<CommentSimple> {
    return this.httpClient.get<CommentSimple>(`${this.commentBaseUri}/${commentId}`)
      .pipe(tap(null, this.errorHandler.handleError('Could not fetch comment')))
  }

  /**
   * Permanently delete a comment
   * 
   * @param commentId id of the comment
   */
  deleteComment(commentId: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.commentBaseUri}/${commentId}`)
      .pipe(tap(null, this.errorHandler.handleError('Could not delete comment')))
  }
}
